File: 02. Algorithms/02. Sorting/04. Quick Sort/quickSort.py
#function that consider last element as pivot, place the pivot at its exact position, and place smaller elements to left of pivot and greater elements to right of pivot.    
def partition (a, start, end):  
    i = (start - 1)  
    pivot = a[end] # pivot element  
      
    for j in range(start, end):  
        # If current element is smaller than or equal to the pivot  
        if (a[j] <= pivot):  
            i = i + 1  
            a[i], a[j] = a[j], a[i]  
      
    a[i+1], a[end] = a[end], a[i+1]  
  
    return (i + 1)  
      
# function to implement quick sort   
def quick(a, start, end): # a[] = array to be sorted, start = Starting index, end = Ending index      
    if (start < end):  
        p = partition(a, start, end) # p is partitioning index  
        quick(a, start, p - 1)  
        quick(a, p + 1, end)  
  
          
def printArr(a): # function to print the array  
    for i in range(len(a)):  
        print (a[i], end = " ")  
  
      
a = [68, 88, 32, 43, 1, 55, 77]  
print("Before sorting array elements are - ")  
printArr(a)  
quick(a, 0, len(a)-1)  
print("\nAfter sorting array elements are - ")  
printArr(a)  

File: 02. Algorithms/02. Sorting/04. Quick Sort/QuickSort.java
public class QuickSort {
    
    public static int sort(int arr[],int si,int ei){

        int p = arr[si];
        int c = 0;
        for(int i = si;i<ei;i++){
            if(p>arr[i]){
                c++;
            }
        }
        int t = arr[c+si];
        arr[c+si] = p;
        arr[si] = t;

        int i = si;
        int j = ei-1;

        while (i<j){

            if(arr[i]<p){
                i++;
            } else if (arr[j]>=p) {
             j--;
            }
            else {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
                j--;
            }
        }

        return si+c;
    }

    public static void quick(int arr[],int si,int ei){

        if(si>=ei){
            return;
        }

        int p = sort(arr,si,ei);
        sort(arr,si,p-1);
        sort(arr,p+1,ei);

    }

    public static void main(String[] args) {

        int arr[] = {2,1,9,6,3};

       quick(arr,0, arr.length);

        for(int i:arr){
            System.out.print(i+" ");
        }
    }
}
